### PR TITLE
feat: add ldk gateway

### PR DESCRIPTION
**Pull request by tvolk131**
Add a third lightning node type to the gateway, a lightning node powered by [`ldk-node`](https://crates.io/crates/ldk-node) which runs in the same process as the gateway. The gateway can be run in ldk mode by setting the following environment variables:

`FM_LDK_ESPLORA_SERVER_URL`: Esplora server where blockchain data is accessed
`FM_LDK_NETWORK`: The bitcoin network being used (e.g. mainnet/testnet)
`FM_PORT_LDK`: The port that the lightning server is running on, where other lightning nodes can access it

The node's data is stored in a subdirectory of the gateway's data folder

This gateway mode should still be considered experimental!!! None of the behavior is tested or has locked-in backwards compatibility yet!

### CODE REVIEW

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:45]: This seems really awkward to me. Does `GatewayLdkClient` need to own this? Can it be passed in as an argument to a function?

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:85]: nit: can avoid string interpolation `error!(?e, "Failed to start LDK Node");`

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:114]: "seed" seems like a weird word for this? This is essentially a "payment interceptor" right? Similar to LND's HTLC interceptor

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:97]: `seed_route_htlcs_stream` only fails in one spot and all you do is log....you can push that log statement into the function and get rid of the error handling here

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:203]: Return error instead of `unwrap`

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:221]: does ldk node not expose this yet?

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:239]: Eventually will be good to implement this. You can create the route hints from the channels that the node is connected to. Not a blocker for this

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:267]: nit: use `fedimint_core::runtime::timeout`

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning/ldk.rs:267]: Can the payment be successful after the 30 second timeout? If so, that could allow an attacker to get paid via lightning and refund their ecash in escrow, stealing money from the gateway

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:266]: What does "routing them" mean here?

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:326]: nit: remove `unwrap`

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:385]: does ldk node support setting the description hash in `receive_for_hash`?

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:495]: We have a very similar function in the CLN extension called `scid_to_u64`. Might be worth to extract somewhere common

Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:495]: I learned that the `lightning` crate exposes an [`scid_from_parts`](https://docs.rs/lightning/0.0.123/lightning/util/scid_utils/fn.scid_from_parts.html) function that does exactly what this function here does, so I removed this one. As for `scid_to_u64`, cln_rpc's `ShortChannelId` struct is represented using a `u64` under the hood, so I made this PR to expose that so we won't need to do any conversion there:

https://github.com/ElementsProject/lightning/pull/7450

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning/ldk.rs]: Perfect

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:46]: The gateway lightning client can be in two states - either it's inactive (not intercepting HTLCs) or active (intercepting HTLCs), and calling `ILnRpcClient::route_htlcs` spins up the HTLC interception, switching from the former state to the latter. Currently these two states are internally represented in the following way:

**Inactive**: `GatewayLdkClient.htlc_stream_seeder_task_handle` is already running and passing incoming HTLCs to the HTLC stream receiver. The HTLC stream receiver is held in `GatewayLdkClient.htlc_stream_receiver_or` until `route_htlcs` is called, at which point this receiver is taken and returned.

**Active**: `GatewayLdkClient.htlc_stream_seeder_task_handle` is still running. But the HTLC stream receiver has been passed to the caller of `route_htlcs` and so `GatewayLdkClient.htlc_stream_receiver_or` is `None`.

What if we changed `GatewayLdkClient.htlc_stream_seeder_task_handle` to be an `Option<tokio::task::JoinHandle<()>>`, removed `GatewayLdkClient.htlc_stream_receiver_or` altogether, and represented the states like this:

**Inactive**: `GatewayLdkClient.htlc_stream_seeder_task_handle_or` is `None` and `ldk-node` events are not being handled yet. The HTLC stream receiver doesn't exist yet.

**Active**: The HTLC stream sender and receiver are created when `route_htlcs` is called. The sender is passed into the htlc stream seeder task, which is started up and assigned to `GatewayLdkClient.htlc_stream_seeder_task_handle_or`. The receiver is returned by `route_htlcs` and never held by `GatewayLdkClient`.

WDYT?

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:269]: Good point... Should we just infinite-loop here and assume that `ldk-node` eventually moves the payment to a non-pending state? (I know this is a pretty gross way of waiting until a payment has been sent, but I believe we need to make some upstream changes to `ldk-node` in order to improve this)

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:115]: Yep, it's a payment interceptor. I renamed it to `handle_next_event()` and renamed `seed_route_htlcs_stream` to `event_handler_task_handle` since if we ever want to handle more `ldk-node` event types, we _must_ do that in this function since it skips over events it doesn't care about.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:98]: Excellent point! Done.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:223]: Not as far as I can see

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:268]: Sorry, that was a strange wording. What I mean is we should only be handling events from `ldk-node`'s event queue in one place to avoid having two threads stomping/skipping over events. Right now, event handling is done in `GatewayLdkClient::event_handler_task_handle`, so we'd need to add functionality there to handle `PaymentSuccessful` and `PaymentFailed` events and somehow be able to tell `GatewayLdkClient::pay` when a payment has succeeded or failed. This can definitely be done, but we'd need some sort of per-paymenthash queue or something.

I cleaned up the comment here

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:387]: I don't think so. Drilling down into the function calls here, we end up at this:

https://github.com/lightningdevkit/rust-lightning/blob/f48a273211d5235e1ff335b2c25f8aacf3642f24/lightning-invoice/src/utils.rs#L431-L453

Where `_create_invoice_from_channelmanager_and_duration_since_epoch()` accepts a `Bolt11InvoiceDescription‎` (which can be a direct description or a hash) but is hardcoded to always be a direct description.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning/ldk.rs:225]: Perhaps add a random identifier in the alias here.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning/ldk.rs:396]: This would probably be good to make an issue upstream

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning/ldk.rs]: Yes I think so. I believe we have similar limitations in LND and CLN currently. I would prefer a payment to be stuck for a long time rather than allowing an attacker to steal funds.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:225]: Done. I just appended the node's pubkey.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs]: Done.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:390]: Created https://github.com/lightningdevkit/ldk-node/issues/325

**Review by dpc — COMMENTED** [review]
LGTM, but someone from @fedimint/lightning should approve.

File: gateway/ln-gateway/src/lightning/mod.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/mod.rs:251]: nit: I think you can specify a default using clap here

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:273]: If you want to implement this, you could potentially use a channel to communicate from the event queue thread to here. Can be done in a follow up

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:313]: This will move `self` rather than clone it, which is what we do in the other implementations. As far as I can tell this won't cause any issues (whereas with LND we need to clone it). I assume this was intentional?

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:337]: nit: "intercepted the payment" not HTLC. Once LNv2 is merged we should refactor all HTLC names out of the codebase

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lightning/ldk.rs:357]: Log an `error!` here, need to have enough info to debug stuff :)

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lightning/ldk.rs:47]: Let's chat about this offline

**Comment by Kodylow** [comment]
Blockers on this, waiting until after 0.4 to merge since we can't get this in this week and need to cut release with the DNS fix. This will be the next big merge.

- Liquidity API and other lightning functionality moved into gateway-cli and gateway api: The only way to interface with the underlying ldk-node is the gateway cli. Normally we use the node's api. For this, we need the gateway-cli to expose api functionality we'd normally just have the node use (open channel, close channel, sweep funds, etc) and add that into the gateway-cli. Instead of only doing this for the ldk node, we're going to implement this generically for all the supported node implementations across the gateway api, which means we can also do lightning side management from the gateway ui in https://github.com/fedimint/ui
  - We already support the liquidity ones for all node impls, we need to add the channel management ones in

- LNv2 requires multi-mint support, but we're dropping that as part of the bolt12 refactor: 0.4 will still use LNv1, but ship with LNv2 binaries that optionally run with an env variable, then for 0.5 both v1 and v2 will run side by side, then 0.6 we retire lnv1.

LNv2 requires an implementation of `pay_private`  for multi-mint payments which requires upstream changes to ldk-node (ldk-node doesn't let us do partial payments of an invoice). 
  - @joschisan : the way to go here is not to require multi-mint payments generally. We have pruned invoices which allow us to do partial payments, which are already supported by LND and CLN, so that made multi-mint payments easy. We intend to drop the pruned-invoices struct because they're causing a bunch of problems elsewhere in the codebase as we prepare for bolt12 support, which will require us to reimplement mutli-mint payments using a different mechanism. So we're going to remove that multi-mint payment as part of LNv2 for now, which removes this blocker.

- LNv2 testing: waiting on ldk to be added to this list of gateway tests):    https://github.com/fedimint/fedimint/pull/5651/files

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lightning/ldk.rs:313]: Yep! This was intentional. The reason that `ILnRpcClient::route_htlcs` consumes `self` and returns `Arc<Self>`  (IIUC) is so that the type checker can enforce that it cannot be called twice. That should still be enforced here.

File: gateway/ln-gateway/src/lightning/ldk.rs
Comment (joschisan) [inline gateway/ln-gateway/src/lightning/ldk.rs:256]: Respecting this parameters is security critical!

**Review by joschisan — APPROVED** [review]

**Pull request merged**
